File: php/src/AddAwsCredentials.php
<?php

namespace HyperSwitchAI\Examples;

require __DIR__ . '/../vendor/autoload.php';

use Dotenv\Dotenv;
use GuzzleHttp\Client;
use GuzzleHttp\Exception\GuzzleException;

// Load environment variables from .env file
$dotenv = Dotenv::createImmutable(__DIR__ . '/..');
$dotenv->load();

const BASE_URL = 'https://api.hyperswitchai.com';
const TOKEN_CACHE_FILE = __DIR__ . '/../.token-cache.json';

function getToken() {
    try {
        // Check if we have a cached token
        if (file_exists(TOKEN_CACHE_FILE)) {
            $cache = json_decode(file_get_contents(TOKEN_CACHE_FILE), true);
            // Check if token is still valid (with 5 minute buffer)
            if ($cache['expiresAt'] > (time() * 1000) + 300000) {
                return $cache['token'];
            }
        }

        // Create HTTP client
        $client = new Client();

        // Get new token
        $response = $client->post(BASE_URL . '/auth', [
            'headers' => [
                'Content-Type' => 'application/json'
            ],
            'json' => [
                'username' => $_ENV['USERNAME'],
                'password' => $_ENV['PASSWORD']
            ]
        ]);

        $data = json_decode($response->getBody(), true);
        $token = $data['token'];

        // Cache token with 1 hour expiry
        file_put_contents(TOKEN_CACHE_FILE, json_encode([
            'token' => $token,
            'expiresAt' => (time() * 1000) + 3600000 // 1 hour
        ]));

        return $token;

    } catch (GuzzleException $e) {
        echo "Auth error: " . $e->getMessage() . "\n";
        throw $e;
    }
}

try {
    $token = getToken();
    
    // Create HTTP client
    $client = new Client();

    // Add AWS credentials
    $response = $client->post(BASE_URL . '/admin/keys/add-key', [
        'headers' => [
            'Content-Type' => 'application/json',
            'Authorization' => 'Bearer ' . $token
        ],
        'json' => [
            'keyId' => 'bedrock-key-1',
            'encryptionFragment' => 'OpenThePodBayDoorsHAL-2001',
            'provider' => 'bedrock',
            'accessKeyId' => 'AKIDFGRTHTHYTHTY',
            'secretAccessKey' => 'ehegruherghekjllkdfjhgkjdfh'
        ]
    ]);

    $statusCode = $response->getStatusCode();

    if ($statusCode === 409) {
        $error = json_decode($response->getBody(), true);
        if (isset($error['error']['code']) && $error['error']['code'] === 'duplicate_key_id') {
            echo "\n\033[33m⚠️  Key already exists with this ID\033[0m\n";
            exit;
        }
    }

    if ($statusCode !== 200) {
        $errorText = $response->getBody();
        echo "\n\033[31mServer response: " . $statusCode . " " . $errorText . "\033[0m\n";
        throw new \Exception("Failed to add key: " . $response->getReasonPhrase());
    }

    echo "\n\033[32m✅ Key added successfully\033[0m\n";

} catch (GuzzleException $e) {
    echo "Error: " . $e->getMessage() . "\n";
} catch (\Exception $e) {
    echo "Error: " . $e->getMessage() . "\n";
} 

File: rust/src/bin/add_aws_credentials.rs
use anyhow::Result;
use colored::Colorize;
use serde::{Deserialize, Serialize};
use std::{env, fs};
use chrono::Utc;

const TOKEN_CACHE_FILE: &str = ".token-cache.json";
const BASE_URL: &str = "https://api.hyperswitchai.com";

#[derive(Debug, Serialize, Deserialize)]
struct TokenCache {
    token: String,
    expires_at: i64,
}

#[derive(Debug, Serialize)]
struct AuthRequest {
    username: String,
    password: String,
}

#[derive(Debug, Deserialize)]
struct AuthResponse {
    token: String,
}

#[derive(Debug, Serialize)]
struct AddAwsKeyRequest {
    #[serde(rename = "keyId")]
    key_id: String,
    #[serde(rename = "encryptionFragment")]
    encryption_fragment: String,
    provider: String,
    #[serde(rename = "accessKeyId")]
    access_key_id: String,
    #[serde(rename = "secretAccessKey")]
    secret_access_key: String,
}

#[derive(Debug, Deserialize)]
struct ErrorResponse {
    error: Option<ErrorDetail>,
}

#[derive(Debug, Deserialize)]
struct ErrorDetail {
    code: String,
}

async fn get_token() -> Result<String> {
    // Check if we have a cached token
    if let Ok(cache_data) = fs::read_to_string(TOKEN_CACHE_FILE) {
        if let Ok(cache) = serde_json::from_str::<TokenCache>(&cache_data) {
            let now = Utc::now().timestamp_millis();
            if cache.expires_at > now + 300000 { // 5 minute buffer
                return Ok(cache.token);
            }
        }
    }

    // Get new token
    let auth_req = AuthRequest {
        username: env::var("USERNAME")?,
        password: env::var("PASSWORD")?,
    };

    let client = reqwest::Client::new();
    let response = client
        .post(format!("{}/auth", BASE_URL))
        .json(&auth_req)
        .send()
        .await?;

    if !response.status().is_success() {
        anyhow::bail!("Authentication failed: {}", response.status());
    }

    let auth_resp: AuthResponse = response.json().await?;

    // Cache token with 1 hour expiry
    let cache = TokenCache {
        token: auth_resp.token.clone(),
        expires_at: Utc::now().timestamp_millis() + 3600000, // 1 hour
    };

    fs::write(
        TOKEN_CACHE_FILE,
        serde_json::to_string_pretty(&cache)?
    )?;

    Ok(auth_resp.token)
}

#[tokio::main]
async fn main() -> Result<()> {
    dotenv::dotenv()?;

    let token = get_token().await?;

    // Add AWS credentials
    let add_key_req = AddAwsKeyRequest {
        key_id: "bedrock-key-1".to_string(),
        encryption_fragment: "OpenThePodBayDoorsHAL-2001".to_string(),
        provider: "bedrock".to_string(),
        access_key_id: "AKIDFGRTHTHYTHTY".to_string(),
        secret_access_key: "ehegruherghekjllkdfjhgkjdfh".to_string(),
    };

    let client = reqwest::Client::new();
    let response = client
        .post(format!("{}/admin/keys/add-key", BASE_URL))
        .header("Content-Type", "application/json")
        .header("Authorization", format!("Bearer {}", token))
        .json(&add_key_req)
        .send()
        .await?;

    if !response.status().is_success() {
        let status = response.status();
        
        if status == reqwest::StatusCode::CONFLICT {
            let error_text = response.text().await?;
            if let Ok(error_resp) = serde_json::from_str::<ErrorResponse>(&error_text) {
                if let Some(error) = error_resp.error {
                    if error.code == "duplicate_key_id" {
                        println!("\n{}", "⚠️  Key already exists with this ID".yellow());
                        return Ok(());
                    }
                }
            }
            println!("\n{} {} {}", "Server response:".red(), status, error_text);
        } else {
            let error_text = response.text().await?;
            println!("\n{} {} {}", "Server response:".red(), status, error_text);
        }
        
        anyhow::bail!("Failed to add key: {}", status);
    }

    println!("\n{}", "✅ Key added successfully".green());

    Ok(())
} 